### ISSUE TITLE
[Feature Request] Support Deep Packet Inspection Bypass/Blocker

### ISSUE THREAD

**Issue by DitFranXX** [AdguardTeam/CoreLibs#1187]
Since some country starts DPI for internet user, I think supports DPI blocker in Adguard(at least Android) is good idea.

Like Android, Many DPI Bypass apps are using VPN tricks, which doesn't compatible with Adguard on non-root devices. 
I'm using Adguard Via Proxy (to Shadowsocks) to solve this, but it has speed decreasing by condition of internet providers' network status.

I know bypassing DPI also has speed decreasing  too. But it doesn't demand on providers' network status. (Which may more faster than VPN)

Since ESNI is still on draft status, DPI blocker is a way of solve this issue without overseas connection.

Ps. May this is not CoreLibs feature request?

Edit: ref to [GoodbyeDPI](https://github.com/ValdikSS/GoodbyeDPI). A DPI blocker.

**Comment by ameshkov** [comment]
Well, we're going to introduce a new product that will help with this. Closing this as dup for now

**Comment by DitFranXX** [comment]
Is that intergrated with current adblock system or whole new product?

**Comment by ameshkov** [comment]
@DitFranXX both, wait a few months plz:)

**Comment by demifiend9** [comment]
Was this new product Adguard VPN? If so, I still think what OP said might be good idea. Using a VPN increases latency and some users don't want their ip to be changed to another location.

For android, Intra app by Google claims to just change DNS by using doH or doT, and using it unblocks every websites blocked by my country's government.
But Adguard's DNS changer doesn't unblock anything.

Intra is open source, perhaps implementing intra's dns changer into adguard android might be good idea?

**Comment by ameshkov** [comment]
> Intra is open source, perhaps implementing intra's dns changer into adguard android might be good idea?

AdGuard can do everything that Intra does + a little bit more.

**Comment by demifiend9** [comment]
It'd be great if that was the case, but it's not.

I'm in India where a lot of websites are blocked.

I use any dns server in Intra: websites are unblocked

Using same dns servers in adguard or android's built in private dns: Websites are not unblocked

I've used various methods to see if dns servers are correctly changed in all these methods. They are, it's just that intra unblocks websites and adguard/android private dns doesn't.

**Comment by demifiend9** [comment]
To add, cloudflare's 1.1.1.1 app also unblocks websites (WITHOUT using warp). Just dns change alone. Using cloudflare over HTTPS or TLS in adguard doesn't unblock.

**Comment by ameshkov** [comment]
Tbh, I have no idea what's going on.

To troubleshoot this issue, we need to get additional logs.

Here's what we need you to do:

1. Open AdGuard -> Settings -> Advanced -> Set "Logging level" to "Record everything". It's desirable to close all background apps that don't concern the problem beforehand.
2. Reproduce the issue and set the logging level back to "Default".
3. Remember the exact time when the issue was reproduced, we will need it to find the corresponding records in the log file.
4. Open AdGuard -> Settings -> Advanced -> Export Logs
5. Additionally, capture the system bug report after reproducing the issue. [Here's the instruction](https://developer.android.com/studio/debug/bug-report).
6. Send both files to `devteam@adguard.com`. Mention the Github issue number and the exact time when the issue was reproduced.

**Comment by demifiend9** [comment]
I have sent the adguard logs. Couldn't take android bug report as they always seemed to fail at 99%.

Anyway, I'm not exactly saying this is a bug in adguard. It's just that intra/cloudflare app seems to do something extra, in a good way. It'd be great for a lot of adguard users in India/South Korea where censorship is a problem if adguard did the same.

**Comment by ameshkov** [comment]
It's just I know how both of these apps do and they literally do nothing else save for redefining DNS. That's why I am asking for the logs.

Btw, I can't see your email in the devteam inbox. Could you please send it directly to me? am aat adguard com

**Comment by shahabhinandan** [comment]
Hi @ameshkov , I had stumbled upon this thread as I faced the same issue myself and was looking for answers.

* Using Intra -> Websites are unblocked
* Using AdGuard (with any server protocol) -> Websites are *not* unblocked
* Using Intra with AdGuard DNS -> Websites unblocked, ads blocked, speed takes a hit.

AdGuard's DNSCrypt or DNS-over-TLS should do what Intra does on paper, but something is missing.

I don't see any reason why Intra is working over AdGuard. Do you have a clue? Any workaround? Also, if the issue is not resolved, shouldn't it be open? Highly appreciate your efforts.

**Comment by demifiend9** [comment]
To add to that, Intra isn't only app that unblocks websites.

XDA recently featured an app called "BraveDNS" which also changes DNS. Article mentioned it's DoH engine is taken from intra and it also unblocks websites.

NextDNS app, Quad9 connect app also unblocks.

It seems like Adguard is in the minority of dns changing apps which isn't able to unblock websites in india. 

I was asked for logs back in February which I definitely sent multiple times to the emails provided. I can again send them if you need.

**Comment by ameshkov** [comment]
> Using AdGuard (with any server protocol) -> Websites are not unblocked

Hmm, I've just got another idea.

What if instead of using full-scale AG, we limit it to being just a DNS filtering app.

Here's how to achieve this:

1. Enable DNS filtering
2. Go to Settings -> Advanced -> Low-level settings -> check `pref.vpn.ipv4.bypass ` and `pref.vpn.ipv6.bypass `
3. Check that DNS queries continue to show up in the filtering log
4. Check if it unblocks domains or not

**Comment by demifiend9** [comment]
Wait I should've double checked before making my last comment.

I just tried right now, Adguard with cloudlfare and Quad9 doh. When I visit any blocked site chrome shows "ERR_CONNECTION_RESET", but when I press reload it successfully opens the website. Sometimes it takes more than one reloads and rarely websites opens up in the first time. I'm pretty sure in the past Adguard just couldn't unblock no matter what I try.

I also the tried ipv4/V6 bypass and it's same same, error at first time and success on a reload.

Anyway, atleast for me consider this problem solved because I don't mind pressing reload button once or twice anytime I'm accessing blocked websites, but I've still send record everything logs to devteam and am in case you want to see them.

**Comment by shahabhinandan** [comment]
I followed your steps as follows:
1. AdGuard is active.
2. DNS Filtering was always enabled with AdGuard DNS (DNS-over-TLS). (As-is)
3. Low level settings -> enabled pref.vpn.ipv4.bypass and pref.vpn.ipv6.bypass
4. Checked blocked website, not working, but showing in [filtering logs](https://imgur.com/a/edfSyj8)
5. Reload doesn't work either.
6. Disabling AG and enabling Intra unblocks the site.

Let me know if I need to share any logs with you. For me reload doesn't work as @demifiend9 stated, but I can confirm that in the past I used to spam the reload button until it worked every once in a while. I don't consider that a solution by any margin, of course.

**Comment by ameshkov** [comment]
Huh, interesting. Most likely this means that among IP addresses returned by the DNS server, only some are blocked.

Maybe Intra checks which ones are alive? Kinda like what AdGuard Home does when "fastest address" option is enabled?

**Comment by shahabhinandan** [comment]
Is there anyway to verify that? Can you check any blocked website from your end? Any other way I can help?

**Comment by ameshkov** [comment]
If there's a website that is blocked and has just one IP address, Intra won't unblock it if I am right.

**Comment by shahabhinandan** [comment]
I don't know any website having only one IP and is blocked in my country, do you? Other way to verify would be to check the same in AG Home and AG Mobile. If one can unblock but the other can't, it will verify the presumption. Also, shall I open a new issue for this since the current one is closed? I have to constantly toggle between Intra and AG right now, since they both give half a pie to me.

**Comment by demifiend9** [comment]
> Is there anyway to verify that? Can you check any blocked website from your end? Any other way I can help?

You could install Adguard home in your pc and use that as your dns server in your phone or another pc with fastest server option enabled and see.

@ameshkov I've tried 15+ websites and have yet to run into something that bravedns/intra couldn't unblock.
Also, does it make sense to install Adguard home and see if it works as I just said? if @shahabhinandan doesn't attempt it then I'll do it later.

**Comment by ameshkov** [comment]
> Also, does it make sense to install Adguard home and see if it works as I just said?

Yep, that'd work.

**Comment by aliafshany** [comment]
> > Using AdGuard (with any server protocol) -> Websites are not unblocked
> 
> Hmm, I've just got another idea.
> 
> What if instead of using full-scale AG, we limit it to being just a DNS filtering app.
> 
> Here's how to achieve this:
> 
> 1. Enable DNS filtering
> 2. Go to Settings -> Advanced -> Low-level settings -> check `pref.vpn.ipv4.bypass ` and `pref.vpn.ipv6.bypass `
> 3. Check that DNS queries continue to show up in the filtering log
> 4. Check if it unblocks domains or not

@ameshkov 
i have this problem in which Intra unblocks many sites that are blocked by DNS manipulation but adguard for android couldn't do that, so i googled it and reached here.
your instruction helped me to bypass those blocked sites partially.
why partially? 
1- because I can't load any pictures on Twitter for example, but Intra does. and also you can tweet images with Intra which with AG is not possible.
2- google safe search is enabled by default from the Gov. side, with intra you can turn it off but with AG it is not possible to do so.

so I'm here to send logs and do your tests in order to make AG fully operational and better.
also, i have a question. How can we do that to the MacOS Adguard version like android?

**Comment by ameshkov** [comment]
Bypassing all the traffic generally limits AdGuard capabilities and limits it to DNS filtering which is no different from just using "private dns" settings in Android. Frankly, I have absolutely no idea how it can help unblock anything:(

**Comment by mhmd21** [comment]
@ameshkov I'm having the same problem as these users. Rethink & Intra apps are able to unblock all blocked sites, but I am unable to do the same thing with Adguard. Is there anything I can do?
